### ISSUE TITLE
Bug in rendering Indic fonts

### ISSUE THREAD

**Issue by rakeshvar** [python-pillow/Pillow#1089]
ImageFont, ImageDraw do not render complex scripts with ligatures correctly. (On Ubuntu 14.04, Python 3.4.0)

This is how it looks in editor gedit:
![gedit](https://cloud.githubusercontent.com/assets/2840906/5931921/5ce383f6-a659-11e4-95a8-9b720cea9a34.png)
This is how Pillow does it.
![pillow](https://cloud.githubusercontent.com/assets/2840906/5931922/5ce3e666-a659-11e4-98cd-5454cb64f4b0.png)

```python
#!/usr/bin/env python3
#-*- coding:utf8 -*-

from PIL import Image, ImageFont, ImageDraw

im = Image.new("RGB",(160, 160))
draw = ImageDraw.Draw(im)

font_telugu = ImageFont.truetype("/usr/share/fonts/truetype/ttf-indic-fonts-core/Pothana2000.ttf",50)
text_telugu = "నిత్య"

font_hindi = ImageFont.truetype("/usr/share/fonts/truetype/ttf-indic-fonts-core/gargi.ttf",50)
text_hindi = "नित्य"

draw.text((10, 10), text_telugu, font=font_telugu)
draw.text((10, 90), text_hindi, font=font_hindi)

im.show()
```

**Comment by wiredfool** [comment]
From a quick run through google, it appears that freetype doesn't support the necessary features to correctly combine glyphs into ligatures for complicated fonts/scripts. There are libraries such as harfbuzz [http://www.freedesktop.org/wiki/Software/HarfBuzz/] that build on freetype (+other dependencies) to render those scripts properly.

**Comment by rakeshvar** [comment]
I have both freetype and harfbuzz up to date on my system.

**Comment by aclark4life** [comment]
I'm going to guess that we have zero harfbuzz integration at present.

**Comment by wiredfool** [comment]
Closing, we'd require harfbuzz for this. We'd welcome pull requests for the integration.

**Comment by wiredfool** [comment]
This should be fixed in 4.2.0 with the merge of #2576

**Comment by rakeshvar** [comment]
Thanks, @ChillarAnand can you please test this?

**Comment by ChillarAnand** [comment]
@wiredfool Thanks for the patch. This doesn't seemed to fixed yet.

```
pip show pillow    
Name: Pillow
Version: 4.2.0
Summary: Python Imaging Library (Fork)
Home-page: https://python-pillow.org
Author: Alex Clark (Fork Author)
Author-email: aclark@aclark.net
License: Standard PIL License
Location: /home/chillar/.virtualenvs/py35/lib/python3.5/site-packages
Requires: olefile
```

Generated image
![p](https://user-images.githubusercontent.com/4463796/27763768-c346ce38-5ea7-11e7-970e-f415141baef1.png)

**Comment by wiredfool** [comment]
Can you check if raqm is in the features list?

```
from PIL import features
print (features.check('raqm'))
```

**Comment by ChillarAnand** [comment]
```
python -c 'from PIL import features;print(features.check("raqm"))'            
False
```

After compiling raqm and successfully installing pillow with 
```
MAX_CONCURRENCY=1 python setup.py build_ext --enable-raqm install
```

it shows
```
python -c 'from PIL import features;print(features.check("raqm"))'            
None
```

**Comment by ChillarAnand** [comment]
@rakeshvar @wiredfool  Its working. Here is the rendered image.

![f](https://user-images.githubusercontent.com/4463796/27784111-9430d47c-5ff6-11e7-929e-e1e1fd623056.png)

**Comment by rakeshvar** [comment]
@ChillarAnand How did you get it to work? My `features.check('raqm')` is showing `False`.

**Comment by ChillarAnand** [comment]
Here is [the script](https://github.com/ChillarAnand/01/blob/master/ubuntu/bin/pillow.sh) I have used to install it. I guess there is some issue with Pothana fonts itself. Its working with Vani and other fonts.

**Comment by menon92** [comment]
@ChillarAnand, 
I have the same problem for `Bangla` language like this
![image](https://user-images.githubusercontent.com/11752205/51304090-69c24500-1a61-11e9-9863-ffd4f709bf5d.png)

How can I solve this

**Comment by ChillarAnand** [comment]
@menon92 Can you try with any other fonts?

**Comment by menon92** [comment]
@ChillarAnand 
I have got the solution https://github.com/python-pillow/Pillow/issues/3593

**Comment by rakeshvar** [comment]
The following worked for me on Ubuntu 18.04 with Python3.6.7:

0. `sudo pip3 install Pillow` -> installs `5.4.1`
1. `sudo apt-get install libfreetype6-dev libharfbuzz-dev libfribidi-dev gtk-doc-tools`
2. Install `libraqm` as described here (using `configure`, `make`, `make install`)
3. `sudo ldconfig` **This step was needed!** 
4. Run the following test script: (Make sure the fonts are installed `sudo apt install fonts-indic`)
```py
from PIL import Image, ImageFont, ImageDraw

im = Image.new("RGB",(160, 160))
draw = ImageDraw.Draw(im)

font_telugu = ImageFont.truetype("/usr/share/fonts/truetype/fonts-telu-extra/Pothana2000.ttf",50)
text_telugu = "నిత్య"

font_hindi = ImageFont.truetype("/usr/share/fonts/truetype/Gargi/Gargi.ttf",50)
text_hindi = "नित्य"

draw.text((10, 10), text_telugu, font=font_telugu)
draw.text((10, 90), text_hindi, font=font_hindi)
im.show()
```

**Comment by khaledhosny** [comment]
Ubuntu 18.04 has `libraqm0`, so you should be able to install it and skip building Raqm manually.

**Comment by HectorsitoXD** [comment]
> Here is [the script](https://github.com/ChillarAnand/01/blob/master/ubuntu/bin/pillow.sh) I have used to install it. I guess there is some issue with Pothana fonts itself. Its working with Vani and other fonts.

Could you find that script for me? I really need it and it's not there. 
```py
from PIL import features
print (features.check('raqm'))
```
is also showing False for me

**Comment by radarhere** [comment]
I can't help you access that exact script, but you can read our documentation at https://pillow.readthedocs.io/en/stable/installation/building-from-source.html

![Screenshot 2024-07-14 at 9 22 27 PM](https://github.com/user-attachments/assets/0ccf048c-4a4e-483c-b4b9-70501269340d)

**Comment by nulano** [comment]
@HectorsitoXD If you'd like more information specific to your system, please open a new issue including your OS and Python version, and how you installed Pillow.

**Comment by HectorsitoXD** [comment]
Ok thanks

**Comment by radarhere** [comment]
https://github.com/python-pillow/Pillow/issues/8232 was created as the new issue.
